### ISSUE TITLE
auto-zoom option doesn't work with stdin input and png output

### ISSUE THREAD

**Issue by wernerdweight** [bbc/audiowaveform#176]
Before I get into the issue itself, let me thank you for this great tool!

**TL;DR:** `stdin` can't be rewound after the data is read from it to get duration.

---

### Bug description:
The zoom option `-z auto, --zoom auto` doesn't work if the input is read from `stdin` and the output is an image.

### Tested version and env:
```
AudioWaveform v1.6.0
Mac OS (12.5.1; tried both a brew-installed version and a compiled one) and Alpine 3.15 (compiled from source)
I tried several audio files (mp3 and WAV) - the behaviour is the same for all of them.
```

### Reproduction steps:
1. Get an audio file (e.g.):
```sh
wget https://file-examples.com/storage/feb2e515cc6339d7ba1ffcd/2017/11/file_example_MP3_2MG.mp3
```
2. Run the following command:
```sh
cat file_example_MP3_2MG.mp3 | audiowaveform --input-format mp3 --output-format png -z auto > test.png
```

### Expected behaviour:
A file named `test.png` is created and contains a waveform of the whole track.
```sh
-rw-r--r--    1 root     root          5475 Oct 11 19:10 test.png
```
The result is expected to be the same as if the following command is used:
```sh
audiowaveform -i file_example_MP3_2MG.mp3 --output-format png -z auto > test.png

Input file: file_example_MP3_2MG.mp3
Calculating audio duration...
Format: Audio MPEG layer III stream
Bit rate: 320000 kbit/s
CRC: no
Mode: normal LR stereo
Emphasis: no
Sample rate: 44100 Hz
Encoding delay: 1105
Padding: 433
Done: 100%
Frames decoded: 2022 (0:52.819)
Duration: 52.7945 seconds
Input file: file_example_MP3_2MG.mp3
Generating waveform data...
Samples per pixel: 2910
Input channels: 2
Output channels: 1
Done: 100%
Frames decoded: 2022 (0:52.819)
Generated 801 points
Image dimensions: 800x250 pixels
Channels: 1
Sample rate: 44100 Hz
Samples per pixel: 2910
Start time: 0 seconds
Start index: 0
Buffer size: 801
Axis labels: yes
Amplitude scale: 1
Output file: (stdout)
Done
```

### Actual behaviour:
A file named `test.png` is created but is empty (zero bytes).
```sh
-rw-r--r--    1 root     root             0 Oct 11 18:56 test.png
```
The output ends abruptly after the duration is determined and the status code is `1`:
```sh
cat file_example_MP3_2MG.mp3 | audiowaveform --input-format mp3 --output-format png -z auto > test.png

Input file: (stdin)
Calculating audio duration...
Format: Audio MPEG layer III stream
Bit rate: 320000 kbit/s
CRC: no
Mode: normal LR stereo
Emphasis: no
Sample rate: 44100 Hz
Encoding delay: 1105
Padding: 433

Frames decoded: 2022 (0:52.819)
Duration: 52.7945 seconds
```

### Assumed cause:
`stdin` can't be rewound as attempted [here](https://github.com/bbc/audiowaveform/blob/abc7c99c963ddbfbd79b3494f30a541c982834b1/src/OptionHandler.cpp#L196-L200): 
```c++
    if (FileUtil::isStdioFilename(input_filename.string().c_str())) {
        if (fseek(stdin, 0, SEEK_SET) != 0) {  // this doesn't work
            return std::make_pair(false, 0);
        }
    }
```

### Assumed solution:
I suppose we should load the data from stdin, store them in memory and then seek on it. I tried looking into this, but I haven't used C++ in ages (and I only used it at university back then, anyway).

**Comment by chrisn** [comment]
Thanks! I'll take a look...

**Comment by chrisn** [comment]
For now, I've just added an error message, as it currently fails silently.

**Comment by wernerdweight** [comment]
Hi @chrisn, I just want to check if this is somewhere on the roadmap. If not, would you potentially be open to having someone else do a PR? As I mentioned, I'm not sure I would be capable of doing that, but I could try persuading someone else to try. 😁  Thanks

**Comment by chrisn** [comment]
Sorry for the delay. Work on this project happens entirely in my own free time.

I have some ideas for how we could solve this, it may be easier if I do it, but then I can't guarantee when that will be. I'd be happy to share ideas if someone else is interested to help.

I'm wondering if either of these workarounds are suitable for you:

```
./audiowaveform --input-filename file_example_MP3_2MG.mp3 --output-format png -z auto > test.png
```
or
```
./audiowaveform --input-format mp3 --output-format png -z auto > test.png < file_example_MP3_2MG.mp3
```

**Comment by wernerdweight** [comment]
> Sorry for the delay. Work on this project happens entirely in my own free time.

No problem, I completely understand.

> I'm wondering if either of these workarounds are suitable for you:

In my case, the audio is received as a byte stream that gets sent to the stdin. I used the `cat file_example_MP3_2MG.mp3` in the bug report above to keep things simple, but in the real-world scenario, there's no audio file (and no storage - it works more like an input/output filter of a sort).

**Comment by chrisn** [comment]
Please try v1.7.0 - this decodes the input audio into memory. Memory usage could be reduced by downsampling the audio as it's read in, then resampling to produce the image, so I might try that approach in a future version.

**Comment by wernerdweight** [comment]
Sorry for the delay. I just did some small testing on my local machine, and it works like a charm! Thank you very much! 👍

**Comment by Grumpier** [comment]
Sorry to jump in after this was closed, but I am experiencing the same problem with v.1.7.0 when streaming to audiowaveform from within a nodejs application. It works when testing via a pipe at the command line, but the following fails only when using --zoom (or -z). As a workaround I have been using pixels-per-second but unfortunately it does not accept a decimal. Any suggestions? Here's the relevant code:

```
 const pngArgs = [
    '--input-filename',    '-',
    '--input-format',      'mp3',
    '--output-filename',   '-',
    '--output-format',     'png',
    '--width',             '400',
    '--height',            '50',
    '--border-color',      '00000000',
    '--background-color',  '00000000',
    '--waveform-color',    '1976d2FF',
    '--axis-label-color',  'FFFFFFFF',
    '-z',                  'auto', 
    '--bits',              '8'
  ]

  const makePng = child_process.spawn("audiowaveform", pngArgs);

  https.get(source, response => {
    response.pipe(makePng.stdin);
    makePng.stdout.pipe(upload(path.parse(track.thumbnailUri).name))
  })
```

**Comment by chrisn** [comment]
Thanks for reporting this. What happens when it fails, do you get an error?

I'm happy to look into it, but I'll need a complete example that I can run.

I guess that your code does something similar to this command line (which works OK), i.e., pipe the MP3 into audiowaveform and pipe the output to a file:

```
cat test.mp3 | audiowaveform --input-filename - --input-format mp3 --output-filename - --output-format png --width 400 --height 50 --bits 8 --z auto | cat > test.png
```

**Comment by Grumpier** [comment]
Thanks for the quick reply. No, there is no thrown error. It simply fails to build the output file. I've run a variety of tests with my code, and the output mechanism is irrelevant. It fails even using straight output to a file rather than via a pipe. I am uploading from a Cloudinary folder and I can confirm that it works okay with the command line:

`curl https://res.cloudinary.com/outlandaudio/video/upload/songs/test.mp3 | audiowaveform --input-filename - --input-format mp3 -o test.png --width 400 --height 50 --bits 8 --z auto`

Assuming you can set up a minimal node server, here's a complete example. If you name this 'testWave.js' just execute it with 'node testWave'. It will fail without throwing an error and not produce 'test.png'. If you comment out --zoom and uncomment --pixels-per-second it will generate a nice little wave. 

Let me know if you need anything else. Thanks for looking into this!

```
const https = require("https");
const child_process = require("child_process");

const source = 'https://res.cloudinary.com/outlandaudio/video/upload/songs/test.mp3';

const pngArgs = [
  '--input-filename',    '-',
  '--input-format',      'mp3',
  '-o',                  'test.png',
  '--width',             '400',
  '--height',            '50',
  '--border-color',      '00000000',
  '--background-color',  '00000000',
  '--waveform-color',    '1976d2FF',
  '--axis-label-color',  'FFFFFFFF',
  '--zoom',               'auto',
  //'--pixels-per-second',  '2',
  '--bits',              '8',
]

const makePng = child_process.spawn("audiowaveform", pngArgs);

https.get(source, response => {
  response.pipe(makePng.stdin);
})

```

**Comment by chrisn** [comment]
OK, here's what I found. You can see the error by making these changes:

```js
https.get(source, response => {
  response.pipe(makePng.stdin);
  makePng.stderr.pipe(process.stderr);
});

makePng.on('exit', (status) => {
  console.log('audiowaveform exited with status', status);
});
```

```
Failed to seek to start of audio
```

It fails because audiowaveform's stdin is a socket. With your `curl` example, stdin is a pipe, not a socket, so the previous change I made for this issue works in that case. It avoids the error by buffering the audio in memory.

I've just pushed a fix that detects when reading from a socket.

**Comment by Grumpier** [comment]
Great. Thanks! Will you be pushing soon to the Homebrew repo so I can try on my Mac?

**Comment by chrisn** [comment]
I have just updated Homebrew, and packages for other systems are coming soon (it takes a while to complete all the builds).

**Comment by Grumpier** [comment]
Works great. Thanks again! 👍
